File: bigquerydatapolicy/v2/python/google-cloud-bigquery-datapolicies/v0/data_policy_service_client_data_policy_get.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# [START bigquerydatapolicy_v2_datapolicyservice_datapolicy_get]
from google.api_core import exceptions
from google.cloud import bigquery_datapolicies_v2


client = bigquery_datapolicies_v2.DataPolicyServiceClient()


def get_data_policy(
    project_id: str,
    location: str,
    data_policy_id: str,
) -> None:
    """
    Retrieves a specific data policy by its resource name.

    This function demonstrates how to fetch the details of an existing data policy
    using its unique identifier within a given project and location.

    Args:
        project_id: The Google Cloud project ID.
        location: The geographic location of the data policy (For example, "us", "eu").
        data_policy_id: The user-assigned ID of the data policy.
    """
    client = bigquery_datapolicies_v2.DataPolicyServiceClient()

    data_policy_name = client.data_policy_path(
        project=project_id,
        location=location,
        data_policy=data_policy_id,
    )

    try:
        response = client.get_data_policy(name=data_policy_name)

        print(f"Successfully retrieved data policy: {response.name}")
        print(f"  Data Policy ID: {response.data_policy_id}")
        print(f"  Data Policy Type: {response.data_policy_type.name}")
        if response.policy_tag:
            print(f"  Policy Tag: {response.policy_tag}")
        if response.grantees:
            print(f"  Grantees: {', '.join(response.grantees)}")
        if response.data_masking_policy:
            masking_policy = response.data_masking_policy
            if masking_policy.predefined_expression:
                print(
                    f"  Data Masking Predefined Expression: {masking_policy.predefined_expression.name}"
                )
            elif masking_policy.routine:
                print(f"  Data Masking Routine: {masking_policy.routine}")

    except exceptions.NotFound:
        print(f"Error: Data policy '{data_policy_name}' not found.")
        print("Make sure the data policy ID, project ID, and location are correct.")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")


# [END bigquerydatapolicy_v2_datapolicyservice_datapolicy_get]

File: bigquerydatapolicy/v2beta1/nodejs/google-cloud-bigquery-datapolicies/data-policy-service-client-data-policy-get.js
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

const process = require('process');

// [START bigquerydatapolicy_v2beta1_datapolicyservice_datapolicy_get]
const {DataPolicyServiceClient} =
  require('@google-cloud/bigquery-datapolicies').v2beta1;
const {status} = require('@grpc/grpc-js');

const client = new DataPolicyServiceClient();

/**
 * Retrieves a specific data policy by its resource name.
 *
 * This sample demonstrates how to get details of an existing data policy.
 * It's useful for verifying the configuration of a data policy or for
 * programmatic access to its properties.
 *
 * @param {string} projectId Your Google Cloud Project ID (e.g., 'my-project-123')
 * @param {string} location The Google Cloud location of the data policy (e.g., 'us')
 * @param {string} dataPolicyId The ID of the data policy to retrieve (e.g., 'my-data-policy')
 */
async function getDataPolicy(
  projectId,
  location = 'us',
  dataPolicyId = 'my-data-policy',
) {
  const name = client.dataPolicyPath(projectId, location, dataPolicyId);

  const request = {
    name: name,
  };

  try {
    const [dataPolicy] = await client.getDataPolicy(request);
    console.log(`Successfully retrieved Data Policy: ${dataPolicy.name}`);
    console.log(`  Data Policy ID: ${dataPolicy.dataPolicyId}`);
    console.log(`  Type: ${dataPolicy.dataPolicyType}`);
    if (dataPolicy.policyTag) {
      console.log(`  Policy Tag: ${dataPolicy.policyTag}`);
    }
    if (dataPolicy.grantees && dataPolicy.grantees.length > 0) {
      console.log(`  Grantees: ${dataPolicy.grantees.join(', ')}`);
    }
    if (dataPolicy.dataMaskingPolicy) {
      console.log(
        `  Data Masking Policy: ${dataPolicy.dataMaskingPolicy.predefinedExpression}`,
      );
    }
  } catch (err) {
    if (err.code === status.NOT_FOUND) {
      console.error(
        `Error: Data Policy '${dataPolicyId}' not found in location '${location}' for project '${projectId}'.`,
      );
      console.error(
        'Ensure the data policy exists and the project/location are correct.',
      );
    } else {
      console.error('Error getting data policy:', err.message);
    }
  }
}
// [END bigquerydatapolicy_v2beta1_datapolicyservice_datapolicy_get]

module.exports = {
  getDataPolicy,
};

